### PR TITLE
Added Some Async Methods in XmlBaseWriter Required by WCF.

### LINKED ISSUE #2502: XmlUTF8TextWriter has not implement WriteStartElementAsync or WriteEndElementAsync
WCF's async operations need to call these methods. See https://github.com/dotnet/wcf/issues/221.

### LINKED ISSUE #2504: Add FlushAsync and WriteBase64Async in XmlDictionaryWriter
WCF needs to use FlushAsync and WriteBase64Async. See https://github.com/dotnet/wcf/issues/222.

### PR DESCRIPTION
**Pull request by shmao**
WCF want to use some of XmlDictionaryWriter's async methods, which were not yet implemented. The PR made these async methods available (in XmlBaseWriter) for WCF to use.

Fixes #2502 
Fixes #2504

### CODE REVIEW

**Comment by shmao** [comment]
@khdang @SGuyGe @mconnew @zhenlan

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs [hunk 1]
```diff
@@ -66,6 +66,14 @@ public override void Flush()
             _writer.Flush();
         }
 
+        public override Task FlushAsync()
+        {
+            if (IsClosed)
+                ThrowClosed();
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Is this exception one that should propagate synchronously, or should it be encapsulated in the Task being returned?

Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Also, it looks like other Task-returning methods check whether there's currently another async operation in flight.  Do these need to do similar checks?

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs [hunk 2]
```diff
@@ -348,6 +356,65 @@ public override void WriteEndAttribute()
             }
         }
 
+        protected override async Task WriteEndAttributeAsync()
+        {
+            if (IsClosed)
+                ThrowClosed();
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Here's an example where the exception is being stored into the returned Task (because this method is async).  We should be consistent, both with these additions and with what's already done in the library.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs [hunk 3]
```diff
@@ -348,6 +356,65 @@ public override void WriteEndAttribute()
...
+
+            if (_writeState != WriteState.Attribute)
+                throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.XmlInvalidWriteState, "WriteEndAttribute", WriteState.ToString())));
+
+            await FlushBase64Async();
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: All of the awaits should be done with ```.ConfigureAwait(false)```.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs [hunk 4]
```diff
@@ -348,6 +356,65 @@ public override void WriteEndAttribute()
             }
         }
 
+        protected override async Task WriteEndAttributeAsync()
+        {
+            if (IsClosed)
+                ThrowClosed();
+
+            if (_writeState != WriteState.Attribute)
+                throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.XmlInvalidWriteState, "WriteEndAttribute", WriteState.ToString())));
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: There's already a ```using``` importing ```System.Runtime.Serialization```.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs [hunk 5]
```diff
@@ -535,6 +682,14 @@ protected void FlushElement()
             }
         }
 
+        protected async Task FlushElementAsync()
+        {
+            if (_writeState == WriteState.Element)
+            {
+                await AutoCompleteAsync(WriteState.Content);
+            }
+        }
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: It'd be more efficient if this were instead implemented like:
```C#
protected Task FlushElementAsync()
{
    return _writestate == WriteState.Element ?
        AutoCompleteAsync(WriteState.Content) :
        Task.CompletedTask;
}
```

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs [hunk 6]
```diff
@@ -1346,6 +1523,71 @@ public override void WriteBase64(byte[] buffer, int offset, int count)
...
+        {
+            if (IsClosed)
+                ThrowClosed();
+
+            if (buffer == null)
+                throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentNullException("buffer"));
+
+            // Not checking upper bound because it will be caught by "count".  This is what XmlTextWriter does.
+            if (offset < 0)
+                throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentOutOfRangeException("offset", SR.Format(SR.ValueMustBeNonNegative)));
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Argument exceptions like these should ideally propagate synchronously out of the call, but here they're getting wrapped in the returned Task.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs [hunk 7]
```diff
@@ -1376,6 +1618,14 @@ private void FlushBase64()
             }
         }
 
+        private async Task FlushBase64Async()
+        {
+            if (_trailByteCount > 0)
+            {
+                await FlushTrailBytesAsync();
+            }
+        }
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: This would be more efficient as:
```C#
private Task FlushBase64Async()
{
    return _trailByteCount > 0 ? 
        FlushTrailBytesAsync() :
        Task.CompletedTask;
}
```

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlNodeWriter.cs [hunk 8]
```diff
@@ -10,25 +10,41 @@
...
 using System.Collections.Generic;
-
+using System.Threading.Tasks;
 
 namespace System.Xml
 {
     internal abstract class XmlNodeWriter
     {
         public abstract void Flush();
+        public virtual Task FlushAsync()
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlNodeWriter.cs]: Why are these virtual rather than abstract?  Since the type is internal, can't we ensure that all derived types override them appropriately?

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs [hunk 9]
```diff
@@ -111,6 +111,30 @@ protected byte[] GetBuffer(int count, out int offset)
             return _buffer;
         }
 
+        protected async Task<Tuple<byte[], int>> GetBufferAsync(int count)
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: I don't know whether it matters for the expected scenarios, but every call to this method is going to require allocating a tuple object; you could avoid that by using a struct type instead.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs [hunk 10]
```diff
@@ -125,18 +149,38 @@ private void EnsureByte()
...
         {
             EnsureByte();
             _buffer[_offset++] = b;
         }
 
+        protected async Task WriteByteAsync(byte b)
+        {
+            await EnsureByteAsync();
+            _buffer[_offset++] = b;
+        }
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: This would be more efficient as something like:
```C#
protected Task WriteByteAsync(byte b)
{
    if (_offset >= bufferLength)
    {
        return FlushBufferAndWriteByteAsync(b);
    }
    else
    {
        _buffer[_offset++] = b;
        return Task.CompletedTask;
    }
}

private async Task FlushBufferAndWriteByteAsync(byte b)
{
    await EnsureByteAsync().ConfigureAwait(false);
    _buffer[_offset++] = b;
}
```

**Comment by stephentoub** [comment]
cc: @krwq

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 1 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: You are right, I should throw an InvalidOperationException synchronously in that case.

Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: It seems that none of XmlBaseWriter's methods checked async operation in flight because the type didn't have any async methods. Does this mean I should add this check in every public sync/async method?

Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: > Does this mean I should add this check in every public sync/async method?

@krwq, what would you recommend here?  I'm not familiar with the expectations around the design for the checking of in-flight operations.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlNodeWriter.cs (hunk 8 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlNodeWriter.cs]: Both XmlBinaryNodeWriter and XmlUTF8NodeWriter inherit from this class. We have no plan to add async operations for XmlBinaryNodeWriter for now.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 7 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Changed. Thanks.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 5 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Fixed. Thanks.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 2 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: I made these exception to be thrown synchronously.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 3 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Added .ConfigureAwait(false) for all awaits.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 4 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Removed.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs (hunk 9 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: Added a new struct for the result type.

**Comment by shmao** [comment]
I updated the PR and added an async check wrapper on XmlDictionaryWriter which does checking of in-flight async operation. Also did changes based on the feedbacks.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs [hunk 11]
```diff
@@ -419,12 +494,27 @@ protected virtual void FlushBuffer()
             }
         }
 
+        protected virtual async Task FlushBufferAsync()
+        {
+            if (_offset != 0)
+            {
+                await _stream.WriteAsync(_buffer, 0, _offset).ConfigureAwait(false);
+                _offset = 0;
+            }
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: You can save some allocations by doing the following:
```csharp
if (_offset != 0)  
{
    var task = _stream.WriteAsync(_buffer, 0, _offset);
    _offset = 0;
    return task;
}
return Task.CompletedTask;
```
You could also think about removing the check for _offset!=0 and presume the caller has done this to avoid a method call/Task allocation unnecessarily. Replace it with a Debug.Assert to catch any code paths which don't do this.

**Comment by shmao** [comment]
Hi, can you please have a look at the pull request? Thank you.

**Comment by stephentoub** [comment]
@krwq, can you please review?

@shmao, who from WCF should review?

**Comment by shmao** [comment]
I had added @mconnew from WCF to review the changes.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 1 above)
Comment (krwq) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: I believe consistency should be our guideline so what we do in other libraries is what we should do (especially System.IO).

From a quick peek in the source code and the previous conversations (I already asked similar question before) I can see what you should do is this:
- if the arguments passed are invalid at the moment of passing (null argument, some negative length etc.) and it will be invalid even when you completely rewrite implementation then you should throw synchronously,
- If there is any slightest chance that it might be invalid only during the runtime (assume implementation may change) then throw asynchronously.

Should you do this in every public sync/async method? Depends: we should consider a risk of doing this change. I believe for the good assessment of compat risk I'd talk to @ericstj.

If the guidelines do not match what we currently have in the code and would make it inconsistent I'd start the design guidelines discussion if we should change everything or follow bad guidelines.

**Comment by shmao** [comment]
@roncain , can you please help review the change? Thanks!

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlBaseWriter.cs]: Basically, yes.  Our guidance has been that any usage error (meaning the developer misused the method, typically by passing incorrect arguments) can/should propagate synchronously; everything else should go out through the returned Task.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs [hunk 12]
```diff
@@ -111,6 +111,30 @@ protected byte[] GetBuffer(int count, out int offset)
...
+            DiagnosticUtility.DebugAssert(count >= 0 && count <= bufferLength, "");
+            int bufferOffset = _offset;
+            if (bufferOffset + count <= bufferLength)
+            {
+                offset = bufferOffset;
+            }
+            else
+            {
+                await FlushBufferAsync().ConfigureAwait(false);
+                offset = 0;
```
Comment (krwq) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: Should you throw here?

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs [hunk 13]
```diff
@@ -131,12 +155,37 @@ protected void WriteByte(byte b)
...
+
         protected void WriteByte(char ch)
         {
             DiagnosticUtility.DebugAssert(ch < 0x80, "");
             WriteByte((byte)ch);
         }
 
+        protected Task WriteByteAsync(char ch)
+        {
+            DiagnosticUtility.DebugAssert(ch < 0x80, "");
```
Comment (krwq) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: Nit: I know it is debug only but I'd add some meaning to the magic number

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextWriter.cs [hunk 14]
```diff
@@ -5,6 +5,7 @@
 using System.Runtime;
```
Comment (krwq) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextWriter.cs]: Why is there a private copy of XML here? Is there something we should add to System.Xml.* to remove it?

Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextWriter.cs]: Sorry, which one is the private copy of XML?

Comment (krwq) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextWriter.cs]: Sorry - disregard my previous message, I thought this was copy of one of System.Xml.ReaderWriter files.

**Comment by krwq** [comment]
LGTM

**Comment by shmao** [comment]
@roncain @iamjasonp @mconnew Can someone from WCF please help review the change? We need this change to unblock 4 WCF issues. Thanks.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs [hunk 15]
```diff
@@ -14,6 +14,8 @@
 using SerializationTypes;
 
 using Xunit;
+using System.Threading.Tasks;
+using System.Threading;
```
Comment (iamjasonp) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: NIT: is this change necessary?

File: src/System.Runtime.Serialization.Xml/tests/XmlDictionaryWriterTest.cs [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+    public class AsyncMemoryStream : MemoryStream
+    {
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            await Task.Delay(10).ConfigureAwait(false);
+            await base.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+    }
+}
```
Comment (iamjasonp) [inline src/System.Runtime.Serialization.Xml/tests/XmlDictionaryWriterTest.cs]: It'd be good to get an additional test to (on a basic level) ensure that the XmlDictionaryAsyncCheckWriter is indeed throwing when async behaviours happen at the same time.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs [hunk 17]
```diff
@@ -0,0 +1,695 @@
...
+            {
+                throw new InvalidOperationException(SR.XmlAsyncIsRunningException);
+            }
+        }
+
+        private Task SetLastTask(Task task)
+        {
+            _lastTask = task;
+            return task;
+        }
```
Comment (iamjasonp) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs]: Other XmlWriters will throw when async operations are called and if XmlWriterSettings.Async is `false`, should we do the same here?

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs [hunk 18]
```diff
@@ -151,12 +200,39 @@ protected void WriteBytes(byte b1, byte b2)
...
+                return Task.CompletedTask;
+            }
+        }
+
+        private async Task FlushAndWriteBytesAsync(byte b1, byte b2)
+        {
+            await FlushBufferAsync().ConfigureAwait(false);
+            _buffer[_offset++] = b1;
+            _buffer[_offset++] = b2;
+        }
```
Comment (iamjasonp) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: Curious why we've done a special case with two bytes?

Comment (iamjasonp) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: Ah, this was hidden in the couple of lines above, there's a sync version of this too.

**Comment by iamjasonp** [comment]
I'd like to see a couple more tests, especially those to test a negative case to know that `CheckAsync()` works as expected, but otherwise LGTM.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs (hunk 17 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs]: Yes, ideally we should do that. But XmlDictionaryWriter.CreateTextWriter does not take an XmlWriterSettings object as parameter, so we cannot set the async value for the writer. Therefore I didn't make the XmlWriterSettings.Async check here. Do you think this is ok?

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs (hunk 18 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlStreamNodeWriter.cs]: Yeah, just wanted to keep the async implementations consistent with the sync version.

File: src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs (hunk 15 above)
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs]: Nice catch. The new tests were added in this file. I forgot to remove these usings when I moved the tests to a new file.

File: src/System.Runtime.Serialization.Xml/tests/XmlDictionaryWriterTest.cs (hunk 16 above)
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/XmlDictionaryWriterTest.cs]: Added a test for the negative case.

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs (hunk 17 above)
Comment (iamjasonp) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlDictionaryAsyncCheckWriter.cs]: Hmm, crap. I don't like it, actually - if we don't do this right the first time, it'll be a compat break to introduce the async switch later if we want to get back in line with the other readers/writers. @roncain, @mconnew, thoughts?

**Comment by iamjasonp** [comment]
I think we need to close on whether or not CheckAsync should look at XmlWriterSettings.Async prior to merging this change.

**Comment by jasonwilliams200OK** [comment]
> I think we need to close on whether or not CheckAsync should look at XmlWriterSettings.Async prior to merging this change. 

@iamjasonp, is the related discussion visible on GitHub? If not, would it be a good idea to merge this as the future PRs can bring prospective changes?

**Comment by iamjasonp** [comment]
Much of the discussion isn't yet GitHub visible as there are implications on the desktop implementation of these APIs too. 

@shmao has been leading the discussion around this one - could you comment on where we are?

**Comment by shmao** [comment]
There were several concerns regarding this PR, e.g. whether to make XmlDictionaryWriter respect XmlWriterSettings.  Now we are done with the discussion on this issue and we decided to take the changes.

 I rebased the commits. I will merge this PR once the new commit passed all checks.

**Comment by stephentoub** [comment]
@dotnet-bot test this please
